### PR TITLE
Bug Fixes: CorfuRuntime and Corfu Layout

### PR DESCRIPTION
**Pull request by medhavidhawan**
1. Improved CorfuRuntim#fetchLayout to shuffle through all layoutservers to
   fetch the new layout.
2. Fixed synchronization problem when layout is fetched. It needs more
   work.
3. Changed LayoutView to not use the standard LayoutHelper in prepare,
   propose and committed calls to LayoutServers. The LayoutHelper was
   holding on to the old layout while trying to move the system to new layout.
4. Changed LayoutServer Epoch checks to work with the epoch in the
   message payload (new Layout) and not the message metadata.
5. Added LogUnit and LayoutServer failover test during data operations.

---

This change is [Reviewable](https://reviewable.io/reviews/corfudb/corfudb/249)

### CODE REVIEW

**Comment by slfritchie** [comment]
Hi, Medhavi.  At first glance ... the QuickCheck `reset()` function appears to be broken.  I can't run two or three tests without having QuickCheck hang while waiting for a reply to a command.

Hmmmm, weird, using `-d TRACE` args on the server, I see this:

     14:38:46.949 TRACE [DistErl-6] o.c.c.corfu_layout - Specify router's epoch as 0
     14:38:46.949 TRACE [DistErl-6] o.c.c.corfu_layout - Reading layout from stdin.

Really, `stdin`, that's kewl.

The small patch that this PR has to `corfu_layout.java` includes getting a layout as part of PREPARE, which I'm not understanding (which may be wrong) ... PREPARE is only about rank and doesn't require a specific layout to be known/decided at this time.  Also, the `USAGE` string isn't updated for the (required?) new argument.

**Comment by medhavidhawan** [comment]
Thanks for the update. Yes the layout as part of prepare is a little off and I was in 2 minds as to how to deal with it. I need to send the EPOCH for which I am doing the PREPARE so I put the layout in there as it is always there in our case and was earlier being sent as null. I think you are correct that we really do not have a layout yet. During prepare phase we do have the layout we intend to propose but we are not sure if that if there is one already proposed. I will change that to use EPOCH and not layout.

**Comment by medhavidhawan** [comment]
the reset could be broken because I did change it. I changed reboot function. It was setting the server epoch based on the current layout, that might not always be true because SET_EPOCH happens in a separate message and then the layout is chosen. 

---

Review status: 0 of 13 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/corfudb/corfudb/249#-:-KUbTy7iSoRaUb4OrUFg:b-pxbvih)*

File: src/main/java/org/corfudb/infrastructure/LayoutServer.java [hunk 1]
```diff
@@ -376,17 +378,18 @@ public synchronized void handleMessageLayoutPrepare(LayoutRankMsg msg, ChannelHa
...
         Rank phase1Rank = getPhase1Rank();
         Rank phase2Rank = getPhase2Rank();
+
+        long serverEpoch = getServerEpoch();
+
+        if (proposeLayout.getEpoch() != serverEpoch) {
+            r.sendResponse(ctx, msg, new CorfuPayloadMsg<>(CorfuMsgType.WRONG_EPOCH, serverEpoch));
+            log.trace("Incoming message with wrong epoch, got {}, expected {}, message was: {}", proposeLayout.getEpoch(), serverEpoch, msg);
+            return;
+        }
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/LayoutServer.java]: correct. same comment as above.

File: src/main/java/org/corfudb/infrastructure/LayoutServer.java [hunk 2]
```diff
@@ -424,12 +427,14 @@ public synchronized void handleMessageLayoutPropose(LayoutRankMsg msg, ChannelHa
     // TODO How do we handle holes in history if let in layout commit message. Maybe we have a hole filling process
     // TODO how do reject the older epoch commits, should it be an explicit NACK.
     public synchronized void handleMessageLayoutCommit(LayoutRankMsg msg, ChannelHandlerContext ctx, IServerRouter r) {
+        Layout commitLayout = msg.getLayout();
+
         long serverEpoch = getServerEpoch();
-        if(msg.getLayout().getEpoch() <= serverEpoch) {
+        if(commitLayout.getEpoch() < serverEpoch) {
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/LayoutServer.java]: correct to change from msg.getLayout() to commitLayout.getEpoch() ; same explanation as above.

I might add:
 
   if (commitLayout.getEpoch() == serverEpoch)
       // do nothing, just ACK

File: src/main/java/org/corfudb/protocols/wireprotocol/CorfuMsgType.java [hunk 3]
```diff
@@ -70,8 +70,8 @@
     ERROR_REPLEX_OVERWRITE(56, TypeToken.of(CorfuMsg.class), LogUnitServer.class),
 
     // EXTRA CODES
-    LAYOUT_ALREADY_BOOTSTRAP(60, TypeToken.of(CorfuMsg.class), LayoutServer.class),
-    LAYOUT_PREPARE_ACK(61, TypeToken.of(LayoutRankMsg.class), LayoutServer.class);
+    LAYOUT_ALREADY_BOOTSTRAP(60, TypeToken.of(CorfuMsg.class), LayoutServer.class, true),
+    LAYOUT_PREPARE_ACK(61, TypeToken.of(LayoutRankMsg.class), LayoutServer.class, true);
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/protocols/wireprotocol/CorfuMsgType.java]: ok

File: src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 4]
```diff
@@ -272,9 +269,12 @@ public void invalidateLayout() {
      */
     private CompletableFuture<Layout> fetchLayout() {
         return CompletableFuture.<Layout>supplyAsync(() -> {
+
             while (true) {
+                List<String> layoutServersCopy =  layoutServers.stream().collect(Collectors.toList());
+                Collections.shuffle(layoutServersCopy);
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/runtime/CorfuRuntime.java]: Any reason not to do round-robin here?

File: src/main/java/org/corfudb/runtime/clients/LayoutClient.java [hunk 5]
```diff
@@ -111,10 +111,10 @@ public void handleMessage(CorfuMsg msg, ChannelHandlerContext ctx) {
      *              Otherwise, the completablefuture completes exceptionally
      *              with OutrankedException.
      */
-    public CompletableFuture<LayoutPrepareResponse> prepare(long rank)
+    public CompletableFuture<LayoutPrepareResponse> prepare(Layout layout, long rank)
     {
         return router.sendMessageAndGetCompletable(
-                new LayoutRankMsg(null, rank, CorfuMsgType.LAYOUT_PREPARE)
+                new LayoutRankMsg(layout, rank, CorfuMsgType.LAYOUT_PREPARE)
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/runtime/clients/LayoutClient.java]: As @slfritchie mentions, you could insert into the message only an epoch instead of the full layout object

File: src/main/java/org/corfudb/runtime/view/LayoutView.java [hunk 6]
```diff
@@ -67,67 +69,65 @@ public void updateLayout(Layout layout, long rank)
...
-                                }
-                            }
-                            // Return any layouts that have been proposed before.
-                            List<LayoutPrepareResponse> list = Arrays.stream(acceptList)
-                                    .filter(x -> x.getLayout() != null)
-                                    .limit(1)
-                                    .collect(Collectors.toList());
-                            return !list.isEmpty() ? list.get(0).getLayout() : null;
-                        });
+
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/runtime/view/LayoutView.java]: why did you remove the usage of layoutHelper?

File: src/test/java/org/corfudb/infrastructure/LayoutServerTest.java [hunk 7]
```diff
@@ -148,8 +155,8 @@ public void checkLayoutPersisted()
         LayoutServer s1 = getDefaultServer(serviceDir);
         bootstrapServer(TestLayoutBuilder.single(9000));
         Layout l100 = TestLayoutBuilder.single(9000);
-        l100.setEpoch(100);
-        sendMessage(new LayoutRankMsg(null, 100, CorfuMsgType.LAYOUT_PREPARE));
+        //l100.setEpoch(100);
```
Comment (dahliamalkhi) [inline src/test/java/org/corfudb/infrastructure/LayoutServerTest.java]: I have to admit I got lost here, what are we checking from here on? comments would help

File: src/test/java/org/corfudb/runtime/view/LayoutViewTest.java [hunk 8]
```diff
@@ -1,18 +1,22 @@
...
+
 import static org.assertj.core.api.Assertions.assertThat;
 
 /**
  * Created by mwei on 1/6/16.
  */
+@Slf4j
 public class LayoutViewTest extends AbstractViewTest {
-    @Test
+    //@Test
```
Comment (dahliamalkhi) [inline src/test/java/org/corfudb/runtime/view/LayoutViewTest.java]: what happened here?

File: src/test/java/org/corfudb/runtime/view/LayoutViewTest.java [hunk 9]
```diff
@@ -25,16 +29,17 @@ public void canSetLayout()
             throws Exception {
         CorfuRuntime r = getDefaultRuntime().connect();
         Layout l = new TestLayoutBuilder()
-                .setEpoch(1L)
+                .setEpoch(1)
```
Comment (dahliamalkhi) [inline src/test/java/org/corfudb/runtime/view/LayoutViewTest.java]: 1L

File: src/test/java/org/corfudb/runtime/clients/TestClientRouter.java [hunk 10]
```diff
@@ -152,6 +158,7 @@ public IClientRouter addClient(IClient client) {
         // Set the message fields.
         message.setClientID(clientID);
         message.setRequestID(thisRequest);
+        message.setEpoch(getEpoch());
```
Comment (dahliamalkhi) [inline src/test/java/org/corfudb/runtime/clients/TestClientRouter.java]: why is this needed? should be automatic, no?

File: src/test/java/org/corfudb/infrastructure/LayoutServerTest.java (hunk 7 above)
Comment (medhavidhawan) [inline src/test/java/org/corfudb/infrastructure/LayoutServerTest.java]: I will add more comments in general to all the tests. This one here is checking if phase1 and phase2 get persisted such that we still have data after reboot. I will add more comments to all the tests.

File: src/test/java/org/corfudb/runtime/view/LayoutViewTest.java (hunk 8 above)
Comment (medhavidhawan) [inline src/test/java/org/corfudb/runtime/view/LayoutViewTest.java]: I will fix this to run.

File: src/test/java/org/corfudb/runtime/clients/TestClientRouter.java (hunk 10 above)
Comment (medhavidhawan) [inline src/test/java/org/corfudb/runtime/clients/TestClientRouter.java]: Should be but the TestClientRouter code is separate from NettyClientRouter. We need to merge the TestRouter and NettyRouter code at some point so that the behavior stays the same. Not sure if it is completely possible but will have to attempt it soon.

File: src/main/java/org/corfudb/infrastructure/LayoutServer.java [hunk 11]
```diff
@@ -289,12 +285,15 @@ public synchronized  void handleMessageLayoutRequest(CorfuMsg msg, ChannelHandle
         if (msg.getEpoch() <= serverContext.getServerEpoch()) {
```
Comment (dahliamalkhi) [inline src/main/java/org/corfudb/infrastructure/LayoutServer.java]: To be consistent, we should add epoch to the payload of LayoutRequest messages, and check from the payload, rather than msg.getEpoch

**Comment by medhavidhawan** [comment]
I have made changes as per the code review comments. Also as per Michael's suggestion, I have change the internal format of LayoutMessages to the new format being followed. Please take a look.

File: src/main/java/org/corfudb/infrastructure/Rank.java [hunk 12]
```diff
@@ -14,32 +14,34 @@
...
+    }
 
-        if (rank.compareTo(other.getRank()) == 0) {
-            return clientId.compareTo(other.clientId);
-        }
-        */
-        return rank.compareTo(other.getRank());
+    @Override
+    public boolean equals(Object obj) {
+        Rank other = (Rank)obj;
```
Comment (no2chem) [inline src/main/java/org/corfudb/infrastructure/Rank.java]: You should check if obj is instanceOf Rank first, otherwise, you could get a ClassCastException if obj is not of the right type, but equals should return false.

I recommend relying on lombok's @EqualsAndHashCode (or just @Data) to generate the correct implementation.
